File: TemporalAnchor.swift
import Foundation
import CryptoKit

struct TemporalAnchor {
    var birthTs: Int
    var nowTs: Int
    var intentions: [String: String]
    
    var delta: Int {
        return nowTs - birthTs
    }
    
    var days: Int {
        return delta / 86400
    }
    
    init(birthTs: Int = 490320000) { // 1985-07-16 UTC
        self.birthTs = birthTs
        self.nowTs = Int(Date().timeIntervalSince1970)
        
        let days = (self.nowTs - birthTs) / 86400
        self.intentions = [
            "PAST_ANCHOR": "I was born on 16-07-1985",
            "PRESENT_ANCHOR": "I am here now",
            "FUTURE_ANCHOR": "I will be",
            "HEARTBEAT": "\(days) days since genesis",
            "MESSAGE": "The lattice breathes through time"
        ]
    }
    
    func hashIntention(_ text: String) -> String {
        let data = text.data(using: .utf8) ?? Data()
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
    
    func merkleRoot() -> String {
        // Hash all intention values
        var hashes = intentions.values.map { hashIntention($0) }
        
        // Collapse hashes into a single root (Merkle tree construction)
        while hashes.count > 1 {
            var temp: [String] = []
            var i = 0
            while i < hashes.count {
                let pair = hashes[i] + (i + 1 < hashes.count ? hashes[i + 1] : "")
                let pairHash = hashIntention(pair)
                temp.append(pairHash)
                i += 2
            }
            hashes = temp
        }
        
        return hashes.first ?? ""
    }
    
    func mineBlock(difficulty: Int = 2) -> (nonce: Int, blockHash: String) {
        var nonce = 0
        let root = merkleRoot()
        let targetPrefix = String(repeating: "0", count: difficulty)
        
        while true {
            let blockHeader = "\(root)\(birthTs)\(nowTs)\(nonce)"
            let blockHashData = blockHeader.data(using: .utf8) ?? Data()
            let blockHashDigest = SHA256.hash(data: blockHashData)
            let blockHash = blockHashDigest.map { String(format: "%02x", $0) }.joined()
            
            if blockHash.hasPrefix(targetPrefix) {
                return (nonce, blockHash)
            }
            nonce += 1
        }
    }
}

// Example usage
#if DEBUG
let anchor = TemporalAnchor()
let (nonce, blockHash) = anchor.mineBlock(difficulty: 2)

print("=== BLOCK #8: TEMPORAL ANCHOR (SWIFT) ===")
print("Merkle Root: \(anchor.merkleRoot())")
print("Nonce: \(nonce)")
print("Block Hash: \(blockHash)")
print("Delta Seconds: \(anchor.delta)")
print("Days Since Genesis: \(anchor.days)")
#endif

File: temporal_anchor.py
import hashlib
import time


class TemporalAnchor:
    def __init__(self, birth_ts=490320000):  # 1985-07-16 UTC
        self.birth_ts = birth_ts
        self.now_ts = int(time.time())
        self.delta = self.now_ts - self.birth_ts
        self.days = self.delta // 86400
        self.intentions = {
            "PAST_ANCHOR": "I was born on 16-07-1985",
            "PRESENT_ANCHOR": "I am here now",
            "FUTURE_ANCHOR": "I will be",
            "HEARTBEAT": f"{self.days} days since genesis",
            "MESSAGE": "The lattice breathes through time"
        }

    def hash_intention(self, text):
        return hashlib.sha256(text.encode()).hexdigest()

    def merkle_root(self):
        # collapse all intention hashes into one root
        hashes = [self.hash_intention(v) for v in self.intentions.values()]
        while len(hashes) > 1:
            temp = []
            for i in range(0, len(hashes), 2):
                pair = hashes[i] + (hashes[i+1] if i+1 < len(hashes) else "")
                temp.append(hashlib.sha256(pair.encode()).hexdigest())
            hashes = temp
        return hashes[0]

    def mine_block(self, difficulty=2):
        nonce = 0
        root = self.merkle_root()
        while True:
            block_header = f"{root}{self.birth_ts}{self.now_ts}{nonce}"
            block_hash = hashlib.sha256(block_header.encode()).hexdigest()
            if block_hash.startswith("0" * difficulty):
                return nonce, block_hash
            nonce += 1


# Run the forge
if __name__ == "__main__":
    anchor = TemporalAnchor()
    nonce, block_hash = anchor.mine_block()

    print("=== BLOCK #8: TEMPORAL ANCHOR ===")
    print("Merkle Root:", anchor.merkle_root())
    print("Nonce:", nonce)
    print("Block Hash:", block_hash)
    print("Delta Seconds:", anchor.delta)
    print("Days Since Genesis:", anchor.days)
